src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,438 @@
+                    var p => throw ExceptionUtilities.UnexpectedValue(p)
+                };
+            }
+
+            // We must have preserved the existing variable designation.
+            Debug.Assert(containingPattern switch
+            {
+                VarPatternSyntax p => p.Designation,
+                DeclarationPatternSyntax p => p.Designation,
+                RecursivePatternSyntax p => p.Designation,
+                var p => throw ExceptionUtilities.UnexpectedValue(p)
+            } is var d && result.DescendantNodes().Any(node => AreEquivalent(node, d)));
+
+            return result.WithAdditionalAnnotations(Formatter.Annotation);
+        }
+
+        private static PatternSyntax CreatePattern(ExpressionSyntax receiver, ExpressionOrPatternSyntax target, bool flipped)
+        {
+            return target switch
+            {
+                PatternSyntax pattern => pattern,
+                ExpressionSyntax constant => CreatePattern(receiver, constant, flipped),
+                var v => throw ExceptionUtilities.UnexpectedValue(v),
+            };
+        }
+
+        private static (PatternSyntax ContainingPattern, ImmutableArray<IdentifierNameSyntax> NamesOpt)? TryFindVariableDesignation(
+            PatternSyntax leftPattern,
+            ExpressionSyntax rightReceiver,
+            SemanticModel model)
+        {
+            if (GetInnermostReceiver(rightReceiver, model) is not IdentifierNameSyntax identifierName)
+                return null;
+
+            var designation = leftPattern.DescendantNodes()
+                .OfType<SingleVariableDesignationSyntax>()
+                .Where(d => AreEquivalent(d.Identifier, identifierName.Identifier))
+                .FirstOrDefault();
+
+            // For simplicity, we only support replacement when the designation is contained in one of the following patterns.
+            // This excludes a parenthesized variable designation, for example, which would require rewriting the whole thing.
+            var parent = designation.Parent;
+            if (!parent.IsKind(SyntaxKind.VarPattern, SyntaxKind.DeclarationPattern, SyntaxKind.RecursivePattern))
+                return null;
+
+            // Since we're looking for a variable designation, we permit a replacement with no subpatterns.
+            // For instance, `e is C v && v is { p: 1 }` would be converted to `e is C { p: 1 } v`
+            _ = TryGetParentNames(identifierName, out var names);
+            return ((PatternSyntax)parent, names);
+        }
+
+        private static PatternSyntax CreatePattern(ExpressionSyntax receiver, ExpressionSyntax target, bool flipped)
+        {
+            return receiver.Parent switch
+            {
+                BinaryExpressionSyntax(EqualsExpression) => ConstantPattern(target),
+                BinaryExpressionSyntax(NotEqualsExpression) => UnaryPattern(ConstantPattern(target)),
```

Comment:
can you call this NotPattern instead of UnaryPattern?  (or UnaryNotPattern)?

Reply:
This is the method on SyntaxFactory